chore: Refactor cell recovery code (#3781)
ethereum/consensus-specs: specs/_features/eip7594/polynomial-commitments-sampling.md

  - [FFTs](#ffts)
    - [`_fft_field`](#_fft_field)
    - [`fft_field`](#fft_field)

>     - [`coset_fft_field`](#coset_fft_field)

Author: Theres a test_fft method, so blocking this until we add a similar method to test `coset_fft`

Author: unblocked with [92d3fcb](https://github.com/ethereum/consensus-specs/pull/3781/commits/92d3fcbc4a4336e6837cc4cc45cffb5b62fee0d9)

        return _fft_field(vals, roots_of_unity)
```

#### `coset_fft_field`

```python
def coset_fft_field(vals: Sequence[BLSFieldElement],
                    roots_of_unity: Sequence[BLSFieldElement],
                    inv: bool=False) -> Sequence[BLSFieldElement]:
    """
    Computes an FFT/IFFT over a coset of the roots of unity. 
    This is useful for when one wants to divide by a polynomial which 
    vanishes on one or more elements in the domain.
    """

>     vals = vals.copy()

Reviewer: If we want, we can delete the extra blank line here.

>     # Multiply each entry in `vals` by succeeding powers of `factor`
>     # ie [vals[0] * factor^0, vals[1] * factor^1, ..., vals[n] * factor^n]
>     def shift_vals(vals: Sequence[BLSFieldElement], factor: BLSFieldElement) -> Sequence[BLSFieldElement]:
>         shift = 1

Reviewer: I think the docstring should be the other style.
```suggestion
    def shift_vals(vals: Sequence[BLSFieldElement], factor: BLSFieldElement) -> Sequence[BLSFieldElement]:
        """
        Multiply each entry in `vals` by succeeding powers of `factor`
        i.e., [vals[0] * factor^0, vals[1] * factor^1, ..., vals[n] * factor^n]
        """
        shift = 1
```

        return _fft_field(vals, roots_of_unity)
```

#### `coset_fft_field`

```python
def coset_fft_field(vals: Sequence[BLSFieldElement], 
                    roots_of_unity: Sequence[BLSFieldElement], 
                    inv: bool=False) -> Sequence[BLSFieldElement]:
    
    # Multiply each entry in `vals` by succeeding powers of `factor`
    # ie [vals[0] * factor^0, vals[1] * factor^1, ..., vals[n] * factor^n]
    def shift_vals(vals: Sequence[BLSFieldElement], factor: BLSFieldElement) -> Sequence[BLSFieldElement]:

>         shift = 1

Reviewer: I'm not opposed to it, but I don't think we typically use nested functions in the specs.

        return _fft_field(vals, roots_of_unity)
```

#### `coset_fft_field`

```python
def coset_fft_field(vals: Sequence[BLSFieldElement], 

>                     roots_of_unity: Sequence[BLSFieldElement], 

Reviewer 2: A bit of trailing whitespace here and along this function in general. Also in the docs of construct_vanishing_polynomial(). Maybe you should ask your editor to highlight them?

Author: Can you elaborate on the issue here? the linter is not showing up any issues

Reviewer: Just an extra space at the end. The linter won't complain about trailing whitespace.

<img width="511" alt="image" src="https://github.com/ethereum/consensus-specs/assets/95511699/6b990c7a-56bb-4828-9200-dc01f5e17e9b">

<img width="511" alt="image" src="https://github.com/ethereum/consensus-specs/assets/95511699/feae1b51-4a6c-4016-96e6-4dd207413b9f">

Reviewer: Here's how you might auto-remove trailing whitespace in vscode:

* https://stackoverflow.com/a/30884298

Author: Gracias

### `construct_vanishing_polynomial`

```python
def construct_vanishing_polynomial(missing_cell_ids: Sequence[CellID]) -> Sequence[BLSFieldElement]:
    """

>     Given the cells that are missing from the data, compute the polynomial that vanishes at every point that

Reviewer: Just a nit, unrelated to your changes.
```suggestion
    Given the cells IDs that are missing from the data, compute the polynomial that vanishes at every point that
```

### `construct_vanishing_polynomial`

```python
def construct_vanishing_polynomial(missing_cell_ids: Sequence[CellID]) -> Sequence[BLSFieldElement]:
    """
    Given the cells IDs that are missing from the data, compute the polynomial that vanishes at every point that
    corresponds to a missing field element.
    
    This method assumes that all of the cells cannot be missing. In this case the vanishing polynomial
    could be computed as Z(x) = x^n - 1, where `n` is FIELD_ELEMENTS_PER_EXT_BLOB.

    We never encounter this case however because this method is used solely for recovery and recovery only

>     works then at most half of the cells are missing.

Reviewer: Something feels wrong the latter part of this sentence:

> and recovery only works then at most half of the cells are missing.

Maybe it should be:

> and recovery only works if half of the cells are available.

Author: I can switch it around to say "at least" instead of "at most"

Reviewer: That would work. But my confusion is with the "works then at" part.

Reviewer: It might also be missing an "if" word. "only works then **if** at most".

### `construct_vanishing_polynomial`

```python
def construct_vanishing_polynomial(missing_cell_ids: Sequence[CellID]) -> Sequence[BLSFieldElement]:
    """
    Given the cells IDs that are missing from the data, compute the polynomial that vanishes at every point that
    corresponds to a missing field element.
    """

    # If all of the cells are missing, then the vanishing polynomial
    # can be computed as Z(x) = x^n - 1, where `n` is FIELD_ELEMENTS_PER_EXT_BLOB
    # 
    # Note: this makes the function complete on all inputs, however this code path should not 

>     # be hit since this method is used for data recovery and if all of the

Reviewer 2: Not a huge fan of this unreachable code. Not gonna die on this hill though. My comments on the thread below.

    for i, coeff in enumerate(short_zero_poly):
        zero_poly_coeff[i * FIELD_ELEMENTS_PER_CELL] = coeff

    return zero_poly_coeff
```

### `recover_data`

```python
def recover_data(cell_ids: Sequence[CellID],
                 cells: Sequence[Cell],
                 zero_poly_coeff: Sequence[BLSFieldElement],
                 roots_of_unity_extended: Sequence[BLSFieldElement]) -> Sequence[BLSFieldElement]:
    """

>     Recover the missing evaluations for the extended blob, given up to half of the evaluations missing.

Reviewer: There's a missing a word at the end ("evaluations missing" vs "evaluations are missing") but I think this could be phrased a little differently for better clarity.
```suggestion
    Recover the missing evaluations for the extended blob, given at least half of the evaluations.
```

    for i, coeff in enumerate(short_zero_poly):
        zero_poly_coeff[i * FIELD_ELEMENTS_PER_CELL] = coeff

    return zero_poly_coeff
```

### `recover_data`

```python
def recover_data(cell_ids: Sequence[CellID],
                 cells: Sequence[Cell],
                 ) -> Sequence[BLSFieldElement]:
    """
    Recover the missing evaluations for the extended blob, given at least half of the evaluations.
    """

    # Get the extended domain. This will be referred to as the FFT domain.
    roots_of_unity_extended = compute_roots_of_unity(FIELD_ELEMENTS_PER_EXT_BLOB)

    # Flatten the cells into evaluations.

>     # If a cell is missing, then its evaluation is zero.

Reviewer: Just a question. Zero is a valid evaluation though. What if all the evals for a cell are zero? Could that happen?

Author: Yep that could happen, if your blob is all zeroes and would be okay

    # Compute Z(x) in monomial form
    # Z(x) is the polynomial which vanishes on all of the evaluations which are missing
    missing_cell_ids = [CellID(cell_id) for cell_id in range(CELLS_PER_EXT_BLOB) if cell_id not in cell_ids]

    zero_poly_coeff = construct_vanishing_polynomial(missing_cell_ids)
    # Convert Z(x) to evaluation form over the FFT domain
    zero_poly_eval = fft_field(zero_poly_coeff, roots_of_unity_extended)
    # Compute (E*Z)(x) = E(x) * Z(x) in evaluation form over the FFT domain
    extended_evaluation_times_zero = [BLSFieldElement(int(a) * int(b) % BLS_MODULUS)
                                      for a, b in zip(zero_poly_eval, extended_evaluation)]
    # Convert (E*Z)(x) to monomial form 
    extended_evaluation_times_zero_coeffs = fft_field(extended_evaluation_times_zero, roots_of_unity_extended, inv=True)

    # Convert (E*Z)(x) to evaluation form over a coset of the FFT domain
    extended_evaluations_over_coset = coset_fft_field(extended_evaluation_times_zero_coeffs, roots_of_unity_extended)

    # Convert Z(x) to evaluation form over a coset of the FFT domain
    zero_poly_over_coset = coset_fft_field(zero_poly_coeff, roots_of_unity_extended)

>     # Compute H(x) = (E*Z)(x) / Z(x) in evaluation form over a coset of the FFT domain

Reviewer 2: H(X) seems to be some new notation introduced here. I tried to use Vitalik's post Q_1, Q_2, Q_3 notation, so that readers can follow. Perhaps it's worth mentioning them in the comments?

Reviewer: We also use this notation in ckzg:

<img width="437" alt="image" src="https://github.com/ethereum/consensus-specs/assets/95511699/36ec2190-b1bc-4689-9238-c653dc4de567">

    for proof_bytes in proofs_bytes:
        assert len(proof_bytes) == BYTES_PER_PROOF

    # Convert cells to coset evals
    cosets_evals = [cell_to_coset_evals(cell) for cell in cells]

    reconstructed_data = recover_data(
        cell_ids,
        cosets_evals

>     )

Reviewer: A bit hard to tell, but this can be joined onto a single line, right?
